library/psa_crypto.c

```diff
@@ -6360,4 +6373,163 @@ psa_status_t psa_crypto_init( void )
     return( status );
 }
 
+psa_status_t psa_pake_setup(
+    psa_pake_operation_t *operation,
+    const psa_pake_cipher_suite_t *cipher_suite )
+{
+    return( psa_driver_wrapper_pake_setup( operation, cipher_suite ) );
+}
+
+psa_status_t psa_pake_set_password_key(
+    psa_pake_operation_t *operation,
+    mbedtls_svc_key_id_t password )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_status_t unlock_status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_slot_t *slot = NULL;
+
+    if( operation->id == 0 )
+        return( PSA_ERROR_BAD_STATE );
+
+    status = psa_get_and_lock_key_slot_with_policy( password, &slot,
+                                                    PSA_KEY_USAGE_DERIVE,
+                                                    PSA_ALG_JPAKE );
+    if( status != PSA_SUCCESS )
+        return( status );
+
+    psa_key_attributes_t attributes = {
+      .core = slot->attr
+    };
+
+    status = psa_driver_wrapper_pake_set_password_key( &attributes, operation,
+                                                       slot->key.data, slot->key.bytes );
+
+    unlock_status = psa_unlock_key_slot( slot );
+
+    return( ( status == PSA_SUCCESS ) ? unlock_status : status );
+}
+
+psa_status_t psa_pake_set_user(
+    psa_pake_operation_t *operation,
+    const uint8_t *user_id,
+    size_t user_id_len )
+{
+    if( operation->id == 0 )
+        return( PSA_ERROR_BAD_STATE );
+
+    if( user_id_len == 0 || user_id == NULL )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
+    return( psa_driver_wrapper_pake_set_user( operation, user_id,
+                                              user_id_len ) );
+}
+
+psa_status_t psa_pake_set_peer(
+    psa_pake_operation_t *operation,
+    const uint8_t *peer_id,
+    size_t peer_id_len )
+{
+    if( operation->id == 0 )
+        return( PSA_ERROR_BAD_STATE );
+
+    if( peer_id_len == 0 || peer_id == NULL )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
+    return( psa_driver_wrapper_pake_set_peer( operation, peer_id,
+                                              peer_id_len ) );
+}
+
+psa_status_t psa_pake_set_role(
+    psa_pake_operation_t *operation,
+    psa_pake_role_t role )
+{
+    if( operation->id == 0 )
+        return( PSA_ERROR_BAD_STATE );
+
+    if( role != PSA_PAKE_ROLE_NONE &&
+        role != PSA_PAKE_ROLE_FIRST &&
+        role != PSA_PAKE_ROLE_SECOND &&
+        role != PSA_PAKE_ROLE_CLIENT &&
+        role != PSA_PAKE_ROLE_SERVER )
+    {
+        return( PSA_ERROR_INVALID_ARGUMENT );
+    }
+
+    return( psa_driver_wrapper_pake_set_role( operation, role ) );
```

Comment:
```
 *   psa_pake_setup()
 *   |
 *   |-- In any order:
 *   |   | psa_pake_set_password_key()
 *   |   | psa_pake_set_user() - not supported
 *   |   | psa_pake_set_peer() - not supported
 *   |   | psa_pake_set_role()
```

So above 4 function should not have driver entry points.
Instead `psa_pake_set_password_key()` and `psa_pake_set_role()` should store `password` and `role` in `psa_pake_operation_t`. Then when `psa_pake_output()` or `psa_pake_input()` is called, the driver should first copy collected `password` and `role` to the corresponding operation context?

Reply:
I think I understand general idea, but not everything is clear to me specially what `acme_pake_setup` should do. Please find below how I see the potential solution:

**1. PAKE operation is divided into 2 stages:**
- collecting inputs stage (first stage)
- computation stage (second stage)

**2. New type to keep inputs set at the collecting inputs stage**

```
struct psa_crypto_driver_pake_inputs_s
{
  psa_algorithm_t MBEDTLS_PRIVATE(alg);
  uint8_t* MBEDTLS_PRIVATE(password);
  size_t MBEDTLS_PRIVATE(password_len);
  uint8_t MBEDTLS_PRIVATE(role);
  psa_key_location_t MBEDTLS_PRIVATE(location);
};

typedef struct psa_crypto_driver_pake_inputs_s
    psa_crypto_driver_pake_inputs_t;
```

**3. Inputs are part of `psa_pake_operation_s`:**
```
struct psa_pake_operation_s
{
    /** Unique ID indicating which driver got assigned to do the
     * operation. Since driver contexts are driver-specific, swapping
     * drivers halfway through the operation is not supported.
     * ID values are auto-generated in psa_crypto_driver_wrappers.h
     * ID value zero means the context is not valid or not assigned to
     * any driver (i.e. none of the driver contexts are active). */
    unsigned int MBEDTLS_PRIVATE(id);
    /* Based on stage (collecting inputs/computation) we select active structure of data union.
     * While switching stage collected inputs are copied to the corresponding operation context (build in, transparent driver, opaque driver) */    
    unsigned int MBEDTLS_PRIVATE(stage);
    union {
        psa_crypto_driver_pake_inputs_t MBEDTLS_PRIVATE(inputs);
        psa_driver_pake_context_t MBEDTLS_PRIVATE(ctx);
    } MBEDTLS_PRIVATE(data);
};
```

**4. Functions behavior**

`psa_pake_setup()`:
- Validate given cipher suit parameters.
- Set `alg` in inputs structure.

`psa_pake_set_password_key()`:
- Validate arguments.
- Set `password`, `password_len`, `key_lifetime` in the inputs structure.

`psa_pake_set_role()`:
- Validate arguments.
- Set `role` in the inputs structure.

`psa_pake_set_user()`:
 - return not supported

`psa_pake_set_peer()`:
 - return not supported

`psa_pake_output()` or `psa_pake_input()`:
- if current `stage` is collecting inputs and all inputs are available:
  - store locally collected inputs,
  - clear pake context (`psa_driver_pake_context_t MBEDTLS_PRIVATE(ctx);`)
  - call `psa_driver_wrapper_pake_setup()`: dispatch based on key lifetime (transparent driver or fallback to build-in impl, opaque driver (not supported)) and set `operation->id` that will be used to dispatch next calls.
  - now when `operation->id` is selected and corresponding context initialized fill it with collected inputs.
  - switch stage to `computation` stage.
  - validate parameters.
  - call `psa_driver_wrapper_pake_input` or `psa_driver_wrapper_pake_output`.

`psa_pake_get_implicit_key()`:
- no special changes

Reply:
Looks good to me

Reply:
Kind of on this topic: I assume that quite a few driver implementers would prefer to perform work on a ‘per-round basis’ because the accelerator libraries they use do so. (That is at least the case for us 😺.) I see the point in having a ‘flexible’ PSA API that fits all/most relevant PAKE schemes, but IMO it wouldn’t hurt if the driver interface came a bit closer to the ‘expected’ / ‘naive’ algorithm APIs. Splitting the input collecting into one stage and then dispatching to the driver definitely seems like the right thing to do; I guess that what I’m thinking is that in an ideal world (as a driver implementer), we could do a similar thing for the driver outputs.

I’m afraid that this would become messy implementation-wise, and that it would increase coupling across the driver interface API, though. The right thing is probably to leave things as is, storing outputs in the driver context until they’re needed is probably not very complicated.